src/data/seq/seq.lean

```diff
@@ -427,14 +436,36 @@ begin
   simp only [zip_with, seq.nth, *]
 end
 
+lemma nth_zip_with (f : α → β → γ) (s : seq α) (t : seq β) (n : ℕ) :
+  nth (zip_with f s t) n = option.bind (nth s n) (λ x, option.map (f x) (nth t n)) :=
+begin
+  cases hx : nth s n with x,
+  { rw [zip_with_nth_none hx, option.none_bind'] },
+  cases hy : nth t n with y,
+  { rw [zip_with_nth_none' hy, option.some_bind', option.map_none'] },
+  { rw [zip_with_nth_some hx hy, option.some_bind', option.map_some'] }
+end
+
 end zip_with
 
 /-- Pair two sequences into a sequence of pairs -/
 def zip : seq α → seq β → seq (α × β) := zip_with prod.mk
 
+lemma nth_zip (s : seq α) (t : seq β) (n : ℕ) :
+  nth (zip s t) n = option.bind (nth s n) (λ x, option.map (prod.mk x) (nth t n)) :=
+nth_zip_with _ _ _ _
+
 /-- Separate a sequence of pairs into two sequences -/
 def unzip (s : seq (α × β)) : seq α × seq β := (map prod.fst s, map prod.snd s)
 
+/-- Enumerate a sequence by tagging each element with its index. -/
+def enum (s : seq α) : seq (ℕ × α) := seq.zip nats s
```

Comment:
I'm curious about the planned usage of this. Also, do you want to define `enum_from` as well? It should make `enum_cons` nicer to state.

Reply:
```
def digits (b n : ℕ) : seq ℕ := of_list (nat.digits b n)

lemma sum_map_enum_digits_eq (b n : ℕ) :
  ((digits b n).enum.map (λ (p : ℕ × ℕ), b ^ p.1 * p.2)).sum = n :=
```